automap.c

```diff
@@ -110,74 +108,304 @@ objects from value lookups), but the hardware-friendly hash table design is what
+    char* p_end = p + dt_size;
+    while (p < p_end && *p != '\0') {
+        p++;
+    }
+    return p;
+}
 
 
-static PyTypeObject AMType;
-static PyTypeObject FAMIType;
-static PyTypeObject FAMVType;
-static PyTypeObject FAMType;
+static inline Py_hash_t
+uint_to_hash(npy_uint64 v) {
+    Py_hash_t hash = (Py_hash_t)(v >> 1); // half unsigned fits in signed
+    if (hash == -1) {
+        return -2;
+    }
+    return hash;
+}
+
+static inline Py_hash_t
+int_to_hash(npy_int64 v) {
+    Py_hash_t hash = (Py_hash_t)v;
+    if (hash == -1) {
+        return -2;
+    }
+    return hash;
+}
+
+
+// This is a adapted from https://github.com/python/cpython/blob/ba65a065cf07a7a9f53be61057a090f7311a5ad7/Python/pyhash.c#L92
+#define HASH_MODULUS (((size_t)1 << 61) - 1)
+#define HASH_BITS 61
+static inline Py_hash_t
+double_to_hash(double v)
+{
+    int e, sign;
+    double m;
+    Py_uhash_t x, y;
+
+    if (isinf(v)) {
+        return v > 0 ? 314159 : -314159;
+    }
+    if (isnan(v)) {
+        return 0;
+    }
+    m = frexp(v, &e);
+    sign = 1;
+    if (m < 0) {
+        sign = -1;
+        m = -m;
+    }
+    x = 0;
+    while (m) {
+        x = ((x << 28) & HASH_MODULUS) | x >> (HASH_BITS - 28);
+        m *= 268435456.0;  /* 2**28 */
+        e -= 28;
+        y = (Py_uhash_t)m;  /* pull out integer part */
+        m -= y;
+        x += y;
+        if (x >= HASH_MODULUS)
+            x -= HASH_MODULUS;
+    }
+    e = e >= 0 ? e % HASH_BITS : HASH_BITS-1-((-1-e) % HASH_BITS);
+    x = ((x << e) & HASH_MODULUS) | x >> (HASH_BITS - e);
+    x = x * sign;
+    if (x == (Py_uhash_t)-1)
+        x = (Py_uhash_t)-2;
+    return (Py_hash_t)x;
+}
+
+
+// This is a "djb2" hash algorithm.
+static inline Py_hash_t
+unicode_to_hash(Py_UCS4 *str, Py_ssize_t len) {
+    Py_UCS4* p = str;
+    Py_UCS4* p_end = str + len;
+    Py_hash_t hash = 5381;
+    while (p < p_end) {
+        hash = ((hash << 5) + hash) + *p++;
+    }
+    if (hash == -1) {
+        return -2;
+    }
+    return hash;
+}
+
+
+static inline Py_hash_t
+string_to_hash(char *str, Py_ssize_t len) {
+    char* p = str;
+    char* p_end = str + len;
+    Py_hash_t hash = 5381;
+    while (p < p_end) {
+        hash = ((hash << 5) + hash) + *p++;
+    }
+    if (hash == -1) {
+        return -2;
+    }
+    return hash;
+}
+
+
+//------------------------------------------------------------------------------
+// the global int_cache is shared among all instances
+
+static PyObject *int_cache = NULL;
+
+
+// NOTE: this used to be a Py_ssize_t, which can be 32 bits on some machines and might easily overflow with a few very large indices. Using an explicit 64-bit int seems safer
+static npy_int64 key_count_global = 0;
+
+
+// Fill the int_cache up to size_needed with PyObject ints; `size` is not the key_count_global.
+static int
+int_cache_fill(Py_ssize_t size_needed)
+{
+    PyObject *item;
+    if (!int_cache) {
+        int_cache = PyList_New(0);
```

Comment:
The main reason is because that would require two loops: one using `PyList_SET_ITEM` for the uncommon case where the list is newly-created, and one using `PyList_Append` for all other resizes. I didn't give it too much thought, and just went with what was simplest.

Reply:
(It looks like this change was copied from the original, with some renamings.)

Reply:
It’s almost certainly faster, but it only happens literally once per process. So I just felt that it wasn’t really worth adding a whole second code path, since there’s no way that it’s a bottleneck (and benchmarking the tiny improvement is super difficult).